feat(portal): symfony sessions

zmilan opened the pull request:
Fixes #9277

#### Short description of what this resolves:
Replace the current session management for the Portal with Symfony Sessions, to iteratively move OpenEMR towards best practice and reduce customization in favor of maintainability.

#### Changes proposed in this pull request:
Switching from PHP native session `$_SESSION` to Symfony Session (with a `NativeSessionStorage`) in `portal/` part of the application

#### Does your code include anything generated by an AI Engine? No

kojiromike commented on portal/account/account.lib.php:
```diff
@@ -510,14 +513,15 @@ function doCredentials($pid, $resetPass = false, $resetPassEmail = ''): bool
 {
-    if (empty($_SESSION['token_id_holder'])) {
+    $session = OpenEMR\Common\Session\SessionUtil::portalSessionStart();
+    if (empty($session->get('token_id_holder'))) {
```
Should we assign this to a name to avoid calling `get` multiple times?

zmilan commented on portal/lib/appsql.class.php:
```diff
@@ -198,14 +199,15 @@ public function portalAudit(?string $type, ?string $rec, array $auditvals, $oelo
         }
 
-        $user_notes .= isset($_SESSION['whereto']) ? (' Module:' . $_SESSION['whereto']) : "";
+        $user_notes .= $session->has('whereto') ? (' Module:' . $session->get('whereto')) : "";
```
It returns `true/false`. This is exact code `return \array_key_exists($name, $this->attributes);`

zmilan replied:
I was wrongly reply about what `has` return... `get` return `null` by default if key doesn't exist with `\array_key_exists` check.

sjpadgett commented:
I'm a little concerned about this be a ginny pig. Portal originally had session management and has management utils as part of Phreeze. Anyway please test this well and good luck because I do support this mission.
I approved the workflows to run.

bradymiller commented on portal/messaging/messages.php:
```diff
@@ -20,25 +20,25 @@
 // Need access to classes, so run autoloader now instead of in globals.php.
```
for example this script can be used by patient portal and core (via their own session/cookies).

zmilan commented on portal/lib/portal_mail.inc.php:
```diff
@@ -286,6 +287,7 @@ function getPortalPatientSentNotes($owner = '', $limit = '', $offset = 0, $searc
  */
 function updatePortalMailMessageStatus($id, $message_status, $owner): void
 {
+    $session = SessionUtil::portalSessionStart();
```
Yes, I noticed that... and that will be the main problem how to handle it properly... I am currently working on cleaning up some things and checking if tests are still working.... and I will move to this problem and using Redis for session (which is also not handled currently for portal)

zmilan replied:
Is there any reliable way to detect from which application is file accessed? For example `portal/messaging/messages.php` file.

zmilan replied:
@bradymiller the problem is not when to start session and how to unify reading from started session (so we can use same code for different sessions). I already achieved that on very simple way with static function in `SessionUtil`... but I will be happy to improve on it and replace it with factory...
The problem is, and it is specific to how the application was built and used, that we have access to the same files directly from different applications. As an example, `portal/messaging/messages.php` is accessed directly from both `core` and `portal` application and there is no application setup/start before we include that file, so that we can decide which session should we start/use... 
I finished my day yesterday with successfully make reading and writing messages from two browsers between administrator and patient by setting up special `Portal = 1` cookie on `portal/index.php` since that is entry point for `portal` application and that helps me to decide if access to `portal/messaging/messages.php`  comes from `core` or `portal`... but I am not very happy with that... I will today try one different idea that comes to me today and see if that can work, which will remove the need for the cookie.

kojiromike replied:
I don't think I fully understand the problem. I think the session name is distinct for the types of applications defined [here](https://github.com/openemr/openemr/blob/e860ab8d94c1bea1d85b8e216cdc742b405eaecf/src/Common/Session/SessionConfigurationBuilder.php#L78): forCore, forOAuth, forApi and forPortal. I take it the session name is not useful here?

The `site_id` is what allows for multi-site deployments of OpenEMR. You get `sites/default` out of the box, and if you want, you can have a completely different OpenEMR deployment at `sites/whatever` if you want. Each site gets its own database connection, document store, and a handful of other things that distinguish it. Basically, it's multi-tenancy via a single deployment.

(So, because each site gets its own database connection, you'd want to be careful about per-site session storage using the database.)

Any given php file that goes through globals.php identifies the site via the `$_SESSION['site_id']`. I do see evidence of that being referenced in /portal as well, so I think you cannot count on it.

Can you try to re-explain the problem? I'm not sure I get it.

zmilan replied:
So, the problem is that I can not detect which application is using shared files (like `messages.php`) and I need to know it because if I start Symfony Session first it spoils previous $_SESSION data for that session name. And as you can see from original code we have `SessionUtil::portalSessionStart();` at the beginning which on my branch starts Symfony Session, so I need somehow to know which application is accessing the file before I initialize SymfonySession, because it could spoil session created with `$_SESSION` for `PortalOpenEMR` name. But if I first check if `$_SESSION` contain some keys I can assume that it is coming from `core` application and if not then it is safe to start Symfony Session...

Other way is to use cookies on login page for each application and detect by them.

kojiromike replied:
In the future, if Symfony Sessions are implemented across everything, would this still be necessary?

zmilan replied:
No, it is only needed at this point

zmilan commented:
@bradymiller, @kojiromike I have issue to replicate locally results of this failing test: https://github.com/openemr/openemr/actions/runs/19061109256/job/54459230540

I am locally running ` openemr-cmd et` and I am getting:
```
OK, but some tests were skipped!
Tests: 145, Assertions: 261, Skipped: 7.
```
Am I doing something wrong?

kojiromike commented:
@zmilan is the log message you're looking for in /var/log/apache2/error.log?

zmilan commented:
> @zmilan is the log message you're looking for in /var/log/apache2/error.log?

I was looking there, but it was empty or showing only warnings, but page was failing with HTTP 500... I will check it more today... thanks

@bradymiller thanks for suggestion, I am still discovering what I can use from `openemr-cmd`, seems very useful :)

zmilan commented:
@kojiromike how do we handle comments on PR? When you open a comment on my PR, who should be responsible for closing it after it is fixed?

zmilan commented:
@kojiromike @sjpadgett do you know how I can start this application expected to be on port 9300?
```PHP
Failed to connect to localhost port 9300 after 0 ms: Could not connect to server for "https://localhost:9300/interface/modules/zend_modules/public/encounterccdadispatch?combination=9&recipient=patient&view=1&me=e317b6244c33dac3c58cf1fc25357d92&site=default
```


```PHP
 /var/www/localhost/htdocs/openemr/ccdaservice/ccda_gateway.php(81): OpenEMR\\Services\\CDADocumentService->portalGenerateCCD()\n#5 {main}\n\nNext Symfony\\Component\\HttpClient\\Exception\\TransportException: Failed to connect to localhost port 9300 after 0 ms: Could not connect to server for "https://localhost:9300/interface/modules/zend_modules/public/encounterccdadispatch?combination=9&recipient=patient&view=1&me=e317b6244c33dac3c58cf1fc25357d92&site=default
```

zmilan commented:
I have come to the realization that on URL: `/portal/patient/onsitedocuments?pid={id}` in "Medical History" form, "Family History" when you click on any of "Diagnosis Code" inputs, you are getting "Site ID is missing from session data!" - and this is on the `master` branch.
I was working on that issue in my branch, and I found where it fails and fixed it, but instead of "Site ID is missing from session data!" I am getting redirected to the core login page... that is when I switched to the `master` branch to double check what should be proper behavior, but I found that it also has the same issue...

Can someone help me understand what should be proper behavior here? 


https://github.com/user-attachments/assets/8c744a39-da53-4166-9538-87068c0ac04d

zmilan commented:
> > Can someone help me understand what should be proper behavior here?
> 
> hi @zmilan pretty sure it should open the code selector at `/interface/patient_file/encounter/select_codes.php` like it does when you add/edit an issue:
> <img alt="image" width="1288" height="871" src="https://private-user-images.githubusercontent.com/5618372/520400967-8d99bfd8-1d94-4883-ae25-b1fe97a5db74.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ0MzIyMDYsIm5iZiI6MTc2NDQzMTkwNiwicGF0aCI6Ii81NjE4MzcyLzUyMDQwMDk2Ny04ZDk5YmZkOC0xZDk0LTQ4ODMtYWUyNS1iMWZlOTdhNWRiNzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMjlUMTU1ODI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2E2MTEyMDBkZTliZTI5NjJkMmViMzhjYzk5ZGZiNzFkNDIzZTU1YjJlM2EyNjgwZWI4NGZlNDBhZWMyMDZiMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.T9co8BzIIjpl36UNm8AuTygylOcChZVXN18Vk4C41JY">

@stephenwaite Thanks... this is helpful... I don't know why it doesn't work on my local when I am on the `master` branch... 
Did you open this from "Portal", or is there another way to access these documents?

stephenwaite commented:
No, that's an example from the main openemr application. Not sure exactly how the portal handled this but @sjpadgett will.

zmilan commented:
> No, that's an example from the main openemr application. Not sure exactly how the portal handled this but @sjpadgett will.

Thanks... I was referring to access from the Portal app: "Clinical Documets" -> "Medical History" -> "Family History" -> click on right-hand input field... For me, that doesn't work even on the `master` branch.

sjpadgett commented:
@zmilan 
The interface/patient_file/encounter/find_code_dynamic.php isn't setup to start a portal session and shouldn't be, so this is an never worked bug.
Patient shouldn't be allowed to add billing codes anyway so the fix is to exclude portal in History layouts and allow patients to only add a text description.
<img width="1920" height="919" alt="image" src="https://github.com/user-attachments/assets/d326ccd2-02a7-4aa7-8cdb-50a6a8b99a00" />

<img width="1920" height="919" alt="image" src="https://github.com/user-attachments/assets/39b6a0f3-f949-498a-86f3-6a1906a325b5" />

zmilan commented:
@sjpadgett Thanks... I will count it as covered from the perspective of my work on the Portal Symfony session

zmilan commented:
@sjpadgett Do you know what is wrong with this:
I am getting this error: `There was an OpenEMR SQL Escaping ERROR of the following string contact_relation.`
on this url (Medical Reports -> View Summary of Care):  /ccdaservice/ccda_gateway.php?action=view&csrf_token_form=2eed185306549c0e5e4427d1d0337dd4eca8a28d
As far as I understand, and following the code, `contact_relation` is not on the white list of tables allowed library/formdata.inc.php function escape_table_name($s) { ... }... now I don't see the `contact_relation` table in the database, so it seems correct... but the question is, do I miss the table, or something else is wrong?

bradymiller commented:
hi @kojiromike and @zmilan , Is this ready for review/testing?

zmilan commented:
I just noticed that `phpstan` complains in a few places, so I will work on cleaning that

kojiromike commented:
@zmilan did you encounter #9940 in working on this?

adunsulag commented on src/Common/Session/SessionWrapperInterface.php:
```diff
@@ -0,0 +1,22 @@
+    public function migrate(bool $destroy = false): bool;
+    public function save(): void;
+    public function clear(): void;
+    public function getSymfonySession(): ?Session;
```
Can we leverage the interface here? Symfony\Component\HttpFoundation\Session\SessionInterface ?

sjpadgett commented on docker/development-easy/docker-compose.yml:
```diff
@@ -90,6 +90,9 @@ services:
       OPENEMR_SETTING_SMTP_Auth: 'TRUE'
       OPENEMR_SETTING_practice_return_email_path: 'noreply@openemr.local'
       OPENEMR_SETTING_Patient_Reminder_Sender_Name: 'OpenEMR Reminders'
+      OPENEMR_PORTAL_JWT_PRIVATE_KEY_FULL_PATH: '/var/www/localhost/htdocs/openemr/sites/default/documents/certificates/portal.pem'
```
I don't think this applies any longer but is it does, example code for creating and handling  JWT in openemr can be found in my ApiExplorer project on my repo.

zmilan replied:
@adunsulag I am sorry for delay, I was sick for a last few days. So, for now, please log in to your PHP container (`openemr/development-easy-openemr-1` in my case), and execute these commands:
`cd openemr/sites/default/documents/certificates/`
`openssl genrsa -out portal.pem 4096` - generate private key
`openssl rsa -pubout -in portal.pem -out portal-public.pem` - generate public key
These command will create keys inside of default folder, and with default names. You can change this and move them to some more secure place.

sjpadgett replied:
@zmilan So I don't have to hunt this down in your commits would you explain the need for the JWT?
Was it for the CCD action?

zmilan replied:
@sjpadgett I am sorry for the delay, but I am still sick and not able to work... As I explained above in one comment: https://github.com/openemr/openemr/pull/9340#issuecomment-3671463066

I found that for "Medical Reports" (summary and customized), we have a "backend" to "backend" call between two applications using HTTPClient (which starts from `src/Services/CDADocumentService.php` Portal App, and calls `interface/modules/zend_modules/public/index.php` which is a separate web app).
We were passing the session id from the Portal app and then initiating a session with that same session id in another web app `session_id($_REQUEST['me']);`. While working on replacing the native PHP session with Symfony, I was not able to replicate this functionality. Symfony Session was just not able to open that session and is timing out waiting. So, the idea was to do things differently, and instead of trying to open the same session, use a JWT token to pass data and authenticate.

sjpadgett replied:
@zmilan I saw the comment but thought you'd comment on my work to remove the need for JWT

zmilan replied:
@sjpadgett I notice that while trying to rebase with latest master branch, but unfortunately I was not able to finish it today… but on the first look it seems that it will not be needed anymore… thanks… I hope I will feel better tomorrow to finish rebasing and retest it with changes…

zmilan replied:
@sjpadgett it seems that the need for the JWT is totally removed after your work on #9950. I have rebased this branch now, but I will need some time today to remove all JWT-related code I added and to test if everything still works as it should. Thanks.

zmilan replied:
This is not needed anymore, and it is removed, together with all JwtService related that I introduced in this PR.

adunsulag commented:
@zmilan let me know when you'd like me to go through this again.

zmilan commented:
@adunsulag I think I will need a little more time for this. For some reason (I am still investigating why), after rebasing when I try to generate a report in "Customized Medical History Report", it lose session after it.

sjpadgett reviewed (commented):
@zmilan 
This seems to be working well at least for the new CCDA actions. However I did find a bug causing actions to throw an exception in `src/Common/ORDataObject/ContactAddress.php`
L-431 when author is null
```
/**
     * @return string
     */
    public function get_author(): string
    {
        return $this->author;
    }

Fix:
/**
     * @return string
     */
    public function get_author(): ?string
    {
        return $this->author;
    }
```
Will continue to test but so far I haven't seen issue of session not persisting on action call returns.

So far I'm liking the change but still learning.

sjpadgett commented:
Duplicated 

> "Customized Medical History Report"

Will take a quick look see!

zmilan commented:
@sjpadgett thanks for checking. I will recheck it and see if that works on this branch too.

zmilan commented:
@sjpadgett I rechecked, and on this branch it doesn't work. I am getting a report, but when I go back to the dashboard and click on anything else, I see that the session is broken.

Additionally, I switched to the master branch to try it there (on the master branch, it works fine), but I see that there is this error as soon as I log in to the portal:
```
An exception has been thrown during the rendering of a template ("xl(): Argument #1 ($constant) must be of type string, null given, called in /var/www/localhost/htdocs/openemr/library/htmlspecialchars.inc.php on line 225") in "portal/partial/cards/_health_snapshot_immunization_card.html.twig" at line 19.
```

I noticed this on this branch, and I already fixed it with this commit: https://github.com/openemr/openemr/pull/9340/changes/64d127d9522f40137c38a723f5af16f2c17e1de3
But I suppose we should fix it separately (and maybe better than what I did) to push it to master faster?

sjpadgett commented:
> ```
> An exception has been thrown during the rendering of a template ("xl(): Argument #1 ($constant) must be of type string, null given, called in /var/www/localhost/htdocs/openemr/library/htmlspecialchars.inc.php on line 225") in "portal/partial/cards/_health_snapshot_immunization_card.html.twig" at line 19.
> ```
> 
> 
>     
>   
> 
> I noticed this on this branch, and I already fixed it with this commit: [64d127d](https://github.com/openemr/openemr/commit/64d127d9522f40137c38a723f5af16f2c17e1de3)

@zmilan I believe @stephenwaite is already addressing this issue or maybe @kojiromike I will pull your branch again and fire up docker and test. Then windows again.

zmilan commented:
@sjpadgett I think I finally know from where it comes.

<img width="1870" height="1032" alt="image" src="https://github.com/user-attachments/assets/f6f2ee0b-b2f9-4a47-ab66-17bcf7130a92" />

```
http://localhost:8300/controller.php?document&retrieve&patient_id=&document_id=2&as_file=false&original_file=false
```
But it only happens, at least on my local data, when I activate "Check All"
<img width="1238" height="406" alt="image" src="https://github.com/user-attachments/assets/cca4baf3-d94b-4cd8-a778-c0e42e9c4cd8" />

And it seems I didn't handle this `controller.php` path, since I didn't notice it before... I can work on fixing it now.

sjpadgett commented:
> But it only happens, at least on my local data, when I activate "Check All"

@zmilan Someone complained about this using our core custom report.
When I catch up a little bit I'll re-pull and retest. Thank you for the quick feedback.

bradymiller commented on interface/modules/zend_modules/public/index.php:
```diff
@@ -16,7 +16,12 @@
-use OpenEMR\Common\Session\SessionUtil;
+use OpenEMR\Common\Logging\SystemLogger;
+use OpenEMR\Common\Session\SessionWrapperFactory;
+use OpenEMR\Common\Auth\JWT\JwtService;
```
Isn't JwtService no longer needed?

zmilan replied:
No, it needs to be removed. I just get busy with some other issue I discovered that somehow was hidden in my previous testing. I will remove it as soon as I complete my current work.

zmilan replied:
After changes from the master branch, it is not needed anymore and it is now removed.

bradymiller commented on interface/patient_file/letter.php:
```diff
@@ -17,21 +17,24 @@
 
 use OpenEMR\Common\Crypto\CryptoGen;
 use OpenEMR\Common\Csrf\CsrfUtils;
+use OpenEMR\Common\Session\SessionWrapperFactory;
```
Just to confirm, the scope of this PR is also conversion to Symfony session in core (since this code is not used by patient portal)?

adunsulag replied:
now that I understand you took the folder approach that makes more sense as I've seen a lot of core file changes.

zmilan commented on src/Common/Auth/JWT/JwtService.php:
```diff
@@ -0,0 +1,347 @@
+<?php
```
@bradymiller it is still not removed since I discovered one more case which needs to be handled. It is not related to this class, but it took me away from it... It will be removed

zmilan replied:
It is removed now. I am sorry for the delay.

kojiromike commented:
@zmilan this needs to wait for #9868, to minimize the phpstan junk, doesn't it?

zmilan commented:
> @zmilan this needs to wait for #9868, to minimize the phpstan junk, doesn't it?

@kojiromike they are not related, but I also think it would be good if #9868 is merged first

adunsulag commented:
@zmilan is this ready to be reviewed again?  I'll go through the code and also do some tests on api / oauth2 to make sure things are working when you're ready.

adunsulag commented on src/Common/Session/SessionUtil.php:
```diff
@@ -119,11 +128,20 @@ public static function setSession($session_key_or_array, $session_value = null):
+            }
         }
         session_write_close();
+        foreach (self::$SESSION_INSTANCES as $symfonySessionInstance) {
```
Marker to come back to this.  How does save work if the session's been closed?

adunsulag commented on src/Common/Session/SessionUtil.php:
```diff
@@ -74,12 +77,18 @@ class SessionUtil
 
     public const DEFAULT_GC_MAXLIFETIME = 14400; // 4 hours
 
+    private static array $SESSION_INSTANCES = [];
```
I'm guessing this is ground work to allow for a core session and a portal session instance to exist simultaneously?   From your code logic I can't see having more than one portal session at a time, correct?

adunsulag reviewed (commented):
Need to step away, but will come back later to this later and go through the portal pieces.  At that point the other stuff should be a straight forward review for any typos / errors and then will do some acceptance testing.

** Edited **
Apparently left out a sentence.  Reviewed through all the Session/* classes and have left the classes in the portal/ folder.

zmilan commented on src/Common/Session/SessionUtil.php:
```diff
@@ -253,4 +310,30 @@ private static function standardSessionCookieDestroy(): void
+            [
+                'expires' => time() + 3600,
+                'path' => '/',
+        //        'domain' => // TODO check how to use proper domain here
```
removed

adunsulag commented on portal/report/portal_custom_report.php:
```diff
@@ -528,8 +530,8 @@ function next_prev(action){
     $titleres = getPatientData($pid, "fname,lname,providerID,DATE_FORMAT(DOB,'%m/%d/%Y') as DOB_TS");
-    if ($_SESSION['pc_facility']) {
-        $sql = "select * from facility where id=" . add_escape_custom($_SESSION['pc_facility']);
+    if ($session->has('pc_facility')) {
```
Need to look at this deeper, as long as pc_facility is never set to 0, there's a slight nuance to this script here.

adunsulag commented on portal/account/register.php:
```diff
@@ -39,18 +42,18 @@
 $data = [
 'global' => $globalsBag->all(),
-'session' => $_SESSION,
+'session' => $_SESSION, // TODO this seems as not used inside of template
```
If its not used we can drop it, otherwise this should probably be $session->all();

zmilan replied:
Done. It seems I forgot it. Thanks.

adunsulag commented on library/sql.inc.php:
```diff
@@ -771,6 +775,7 @@ function sqlRollbackTrans(): void
  */
 function getPrivDB()
 {
+    global $session;
```
Why the global?  Why not grab it from the singleton? 

$session = SessionWrapperFactory::getInstance()->getWrapper();

zmilan replied:
I replaced all `global` usage of the `$session` with proper `SessionWrapperFactory::getInstance()->getWrapper();`

adunsulag commented on library/patient.inc.php:
```diff
@@ -133,7 +135,7 @@ function getInsuranceProvidersExtra()
 function getFacility($facid = 0)
 {
-    global $facilityService;
+    global $facilityService, $session;
```
Seems like it would be better to use the singleton methods here instead of the global.  That way nothing in code can redefine the $session variable.

adunsulag replied:
Note there are several other references in this file where the same thing is done.

zmilan replied:
Replaced with a singleton. Thanks

adunsulag commented on library/forms.inc.php:
```diff
@@ -135,9 +137,10 @@ function getFormNameByFormdir($formdir)
 
 function getDocumentsByEncounter($patientID = null, $encounterID = null)
 {
+    global $session;
```
Seems like it would be better to use the singleton methods here instead of the global.  That way nothing in code can redefine the $session variable.

adunsulag commented on library/auth.inc.php:
```diff
@@ -127,8 +131,8 @@ function authCloseSession(): void
   // login will default to that same site.
-    global $incoming_site_id;
-    $incoming_site_id = $_SESSION['site_id'] ?? '';
+    global $incoming_site_id, $session;
```
Seems like it would be better to use the singleton methods here instead of the global.  That way nothing in code can redefine the $session variable.

adunsulag commented on library/api.inc.php:
```diff
@@ -37,9 +40,9 @@ function formFooter(): void
 function formSubmit($tableName, $values, $id, $authorized = "0")
 {
-    global $attendant_type;
+    global $attendant_type, $session;
```
Seems like it would be better to use the singleton methods here instead of the global.  That way nothing in code can redefine the $session variable.

adunsulag commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -486,6 +489,7 @@ function generate_receipt($patient_id, $encounter = 0): void
     global $num_optional_columns, $rcpt_num_method_columns, $rcpt_num_ref_columns, $rcpt_num_amount_columns;
     global $TAXES_AFTER_ADJUSTMENT;
     global $facilityService, $alertmsg;
+    global $session;
```
Seems like it would be better to use the singleton methods here instead of the global.  That way nothing in code can redefine the $session variable.

adunsulag commented on interface/patient_file/merge_patients.php:
```diff
@@ -287,10 +290,11 @@ function mergeRows($tblname, $colname, $source_pid, $target_pid): void
          */
         function logMergeEvent($target_pid, $event_type, $log_message): void
         {
+            global $session;
```
use singleton method to grab session.

adunsulag replied:
there's two places in this file this is done.

adunsulag reviewed (commented):
Got through all the classes but globals.php and C_Documents.class.php.  Will review those later and then do manual testing on UX where we don't have any e2e tests.

zmilan commented on interface/orders/single_order_results.inc.php:
```diff
@@ -315,7 +318,7 @@ function generate_result_row(&$ctx, &$row, &$rrow, $priors_omitted = false): voi
 function generate_order_report($orderid, $input_form = false, $genstyles = true, $finals_only = false)
 {
-    global $aNotes;
+    global $aNotes, $session;
```
Done. Thanks.

zmilan commented on interface/login/login.php:
```diff
@@ -29,21 +29,30 @@
+require_once(__DIR__ . "/../../vendor/autoload.php");
+
+$globalsBag = OEGlobalsBag::getInstance();
+$globalsBag->set('already_autoloaded', true);
```
Removed. Thanks

zmilan commented on portal/portal_payment.php:
```diff
@@ -71,10 +72,10 @@
     "LEFT JOIN patient_access_onsite pao ON pao.pid = patient_data.pid " .
     "WHERE patient_data.pid = ? AND pao.portal_pwd_status = 1";
 $portalPatient = sqlQueryNoLog($query, $pid);
```
Thanks @kojiromike ... it is fixed now. I also noticed one more case a few lines down: https://github.com/openemr/openemr/pull/9340/changes/8a14eb84054408a15c2c3b2774bc0830976c606c

bradymiller commented on controller.php:
```diff
@@ -1,5 +1,16 @@
+
+    $ignoreAuth_onsite_portal = true; // ignore the standard authentication for a regular OpenEMR user
+}
+
```
hi @zmilan , Opening up entire controller to a patient portal session is a bit scary. Is there a way to only allow this for specific requests? (sounds like just for document requests, which already have a nice safeguard to prevent patient's from getting another patient's documents; so if just do above for that type of request, should be ok)

zmilan replied:
Ok, but it is still not clear for me what I should do: a) leave it as it is now, b) revert my changes in the `controller.php`.

zmilan commented on library/clinical_rules.php:
```diff
@@ -22,8 +22,11 @@
+use OpenEMR\Common\Session\SessionWrapperFactory;
 use OpenEMR\Services\FacilityService;
 
+$session = SessionWrapperFactory::getInstance()->getWrapper();
```
Yes, that is correct. I don't know why my PHPStorm doesn't show it as not in use... It is left from removing the usage of `global`.

bradymiller commented on src/Common/Session/PHPSessionWrapper.php:
```diff
@@ -0,0 +1,88 @@
+        $webroot = $globalsBag->get('webroot');
+        if ($webroot !== null && session_status() !== PHP_SESSION_ACTIVE) {
+            SessionUtil::coreSessionStart($webroot, false);
+        }
```
yep, would keep it as it is

bradymiller commented:
nice stuff!
